### PR TITLE
Reworking on PR #745

### PR DESCRIPTION
**Pull request by ayende**
Make it easier to review: https://github.com/ravendb/ravendb/pull/745

### CODE REVIEW

File: src/Raven.Client/DocumentStoreBase.cs [hunk 1]
```diff
@@ -34,6 +34,35 @@ protected DocumentStoreBase()
...
+        {
+            BeforeStoreEvent?.Invoke(session, id, entityinstance, metadata);
+        }
+
+        public virtual void OnAfterStore(Documents.InMemoryDocumentSessionOperations session, string id, object entityinstance, IDictionary<string, string> metadata)
+        {
+            AfterStoreEvent?.Invoke(session, id, entityinstance, metadata);
+        }
+
+        public bool BeforeStoreEmpty => (BeforeStoreEvent == null);
```
Comment (ayende) [inline src/Raven.Client/DocumentStoreBase.cs:47]: Why do you need those?

File: src/Raven.Client/Documents/Async/AsyncDocumentSession.cs [hunk 2]
```diff
@@ -304,45 +304,62 @@ protected override Task<string> GenerateKeyAsync(object entity)
             return Conventions.GenerateDocumentKeyAsync(DatabaseName, AsyncDatabaseCommands, entity);
         }
 
-        public async Task<BlittableJsonReaderObject> GetMetadataForAsync<T>(T instance)
+        private readonly List<object> _entitiesWithMetadataInstance = new List<object>();
```
Comment (ayende) [inline src/Raven.Client/Documents/Async/AsyncDocumentSession.cs:307]: This needs to be removed, there is not need of it

File: src/Raven.Client/Documents/InMemoryDocumentSessionOperations.cs [hunk 3]
```diff
@@ -842,16 +780,15 @@ private void PrepareForEntitiesDeletion(SaveChangesData result, IDictionary<stri
...
                     {
                         etag = documentInfo.ETag;
+
                         if (documentInfo.Entity != null)
                         {
-                            foreach (var deleteListener in TheListeners.DeleteListeners)
-                            {
-                                deleteListener.BeforeDelete(documentInfo.Id, documentInfo.Entity, documentInfo.Metadata);
-                            }
+                            _documentStore.OnAfterStore(this, documentInfo.Id, documentInfo.Entity, GetMetadataFor(documentInfo.Entity));
```
Comment (ayende) [inline src/Raven.Client/Documents/InMemoryDocumentSessionOperations.cs:786]: Here you are doing work that you probably won't need if no one is lisetning

File: src/Raven.Client/Documents/Query/AbstractDocumentQuery.cs [hunk 4]
```diff
@@ -797,20 +788,12 @@ public virtual Lazy<int> CountLazily()
             if (queryOperation != null)
                 return queryOperation;
             ClearSortHints(AsyncDatabaseCommands);
-            ExecuteBeforeQueryListeners();
+            theSession.OnBeforeQueryExecuted(this);
```
Comment (ayende) [inline src/Raven.Client/Documents/Query/AbstractDocumentQuery.cs:791]: This also need to get the session as well

File: src/Raven.Client/Documents/SessionOperations/BatchOperation.cs [hunk 5]
```diff
@@ -85,13 +85,8 @@ public void SetResult(BatchResult result)
...
                 }
-                //TODO - Efrat
-                /*GenerateEntityIdOnTheClient.TrySetIdentity(entity, batchResult.Key);*/
-                //TODO - Efrat - after listeners
-                foreach (var documentStoreListener in _session.Listeners.StoreListeners)
-                {
-                    documentStoreListener.AfterStore(key, entity, metadata);
-                }
+
+                _session.OnAfterStore(_session, key, entity, _session.GetMetadataFor(entity));
```
Comment (ayende) [inline src/Raven.Client/Documents/SessionOperations/BatchOperation.cs:89]: Doing work that you probably won't need here as well

File: src/Raven.Client/IDocumentStore.cs [hunk 6]
```diff
@@ -18,11 +18,42 @@
...
+    public delegate void AfterStoreDelegate(Documents.InMemoryDocumentSessionOperations session, string id, object entityInstance, IDictionary<string, string> metadata);
+    /// <summary>
+    /// Delete delegate
+    /// </summary>
+    public delegate void BeforeDeleteDelegate(Documents.InMemoryDocumentSessionOperations session, string id, object entityInstance, IDictionary<string, string> metadata);
+
+    /// <summary>
+    /// Query delegate
+    /// </summary>
+    public delegate void BeforeQueryExecutedDelegate(Documents.IDocumentQueryCustomization queryCustomization);
```
Comment (ayende) [inline src/Raven.Client/IDocumentStore.cs:33]: Need to know on what session it operates here

File: src/Raven.Client/DocumentStoreBase.cs (hunk 1 above)
Comment (efratshenhar) [reply src/Raven.Client/DocumentStoreBase.cs:47]: I can check if BeforeStoreEvent  is null Only in DocumentStoreBase

Comment (ayende) [reply src/Raven.Client/DocumentStoreBase.cs:47]: This is pretty much only relevant if you need to do some amount of work before calling the event, so it isn't worth it if there are no subscribers

**Pull request merged**